New extension: GELF logging support (#5694)
quarkusio/quarkus: docs/src/main/asciidoc/central-logging.adoc

////
This guide is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
////
= Quarkus - Central log management (Graylog, Logstash, Fluentd)

include::./attributes.adoc[]

> This guide explains how you can send you log to a central logging system like Graylog, Logstash (inside the Elastic Stack or ELK - Elasticsearch, Logstash, Kibana) or

Reviewer: ```suggestion
This guide explains how you can send your logs to a central logging system like Graylog, Logstash (inside the Elastic Stack or ELK - Elasticsearch, Logstash, Kibana) or
```

Fluentd (inside EFK - Elasticsearch, Fluentd, Kibana).

> There is a lot of different way to centralize your logs (if you're using Kubernetes, the simplest way is to log to the console and ask you cluster administrator to integrate a central log manager inside your cluster).

Reviewer: ```suggestion
There are a lot of different ways to centralize your logs (if you are using Kubernetes, the simplest way is to log to the console and ask you cluster administrator to integrate a central log manager inside your cluster).
```

Author: done

> In this guide, we will expose how to send them to an external tool using the `quarkus-logging-gelf` extension that can use TCP or UDP to send logs in the Graylog Extented Log Format (GELF).

Reviewer: ```suggestion
In this guide, we will expose how to send them to an external tool using the `quarkus-logging-gelf` extension that can use TCP or UDP to send logs in the Graylog Extended Log Format (GELF).
```

The `quarkus-logging-gelf` extension will add a GELF log handler to the underlying logging backend that Quarkus uses (jboss-logmanager).
By default, it is disabled, if you enable it but still use another handler (by default the console handler is enabled), your logs will be sent to both handlers.

== Example application

> The following examples will all be based on the same example application that you can create based on the following steps:

Reviewer: ```suggestion
The following examples will all be based on the same example application that you can create with the following steps:
```

- Create an application with the `quarkus-logging-gelf` extension. You can use the following Maven command to create it:

> ```

Reviewer: Please use proper Asciidoc source for all the source extracts.

Author: Done, sorry for this, markdown habits :)

mvn io.quarkus:quarkus-maven-plugin:1.0.0.Final:create \
    -DprojectGroupId=org.acme \
    -DprojectArtifactId=gelf-logging \
    -DclassName="org.acme.quickstart.GelfLoggingResource" \
    -Dpath="/gelf-logging" \
    -Dextensions="logging-gelf"
```

> - For demonstration purpose, we create an endpoint that do nothing but log a sentence. You don't need to do this inside your application.

Reviewer: ```suggestion
- For demonstration purpose, we create an endpoint that does nothing but log a sentence. You don't need to do this inside your application.
```

> ```

Reviewer: Same here about using source + you need to say it's Java.

Author: done

> @Path("/gelf-logging")

Reviewer: Please include the imports. in the source file. Otherwise people have a hard time reproducing the steps.

Author: done

@ApplicationScoped
public class GelfLoggingResource {
    private static final Logger LOG = Logger.getLogger(GelfLogHandlerResource.class);

    @GET
    public void log() {
        LOG.info("Some useful log message");
    }

}
```

> - Configure the gelf log handler to send logs to an external UDP endpoint on the port 12201

Reviewer: ```suggestion
- Configure the GELF log handler to send logs to an external UDP endpoint on the port 12201:
```

```
quarkus.log.handler.gelf.enabled=true
quarkus.log.handler.gelf.host=localhost
quarkus.log.handler.gelf.port=12201
```

== Send logs to Graylog

To send logs to Graylog, you first need to launch the components that compose the Graylog stack:

- MongoDB
- Elasticsearch
- Graylog

You can do this via the following docker-compose file that you can launch via `docker-compose run -d`:

```
version: '3.2'

services:
  elasticsearch:

>     image: docker.elastic.co/elasticsearch/elasticsearch-oss:6.8.2

Reviewer: We can't use a more recent version?

Author: This is the version used inside the Graylog documentation: http://docs.graylog.org/en/3.1/pages/installation/docker.html#here
A more recent version should works but I will need to re-run all the test ...

Author: In fact Graylog is only compatible with Elasticsearch 5.6 and 6.8: https://docs.graylog.org/en/3.1/pages/configuration/elasticsearch.html

The other examples should work on more recent version of Elasticsearch but to keep it simple I prefere to use the same for all

    networks:
      - graylog

  mongo:
    image: mongo:4.0
    networks:
      - graylog

  graylog:
    image: graylog/graylog:3.1
    ports:
      - "9000:9000"
      - "12201:12201/udp"
      - "1514:1514"
    environment:
      GRAYLOG_HTTP_EXTERNAL_URI: "http://127.0.0.1:9000/"
    networks:
      - graylog
    depends_on:
      - elasticsearch
      - mongo

networks:
  graylog:
    driver: bridge
```

Then, you need to create a UDP input in Graylog.
You can do it from the Graylog web console (System -> Input -> Select GELF UDP) or via the API.

This curl example will create a new Input of type GELF UDP, it uses the default login from Graylog (admin/admin).

```
curl -H "Content-Type: application/json" -H "Authorization: Basic YWRtaW46YWRtaW4=" -H "X-Requested-By: curl" -X POST -v -d \
'{"title":"udp input","configuration":{"recv_buffer_size":262144,"bind_address":"0.0.0.0","port":12201,"decompress_size_limit":8388608},"type":"org.graylog2.inputs.gelf.udp.GELFUDPInput","global":true}' \
http://localhost:9000/api/system/inputs
```

Launch your application, you should see your logs arriving inside Graylog.

== Send logs to Logstash / the Elastic Stack (ELK)

> Logstash comes by default with an Input plugin that can understand the GELF format, we will first create a pipeline that enable this plugin.

Reviewer: ```suggestion
Logstash comes by default with an Input plugin that can understand the GELF format, we will first create a pipeline that enables this plugin.
```

Create the following file  in `$HOME/pipelines/gelf.conf`:
```
input {
  gelf {
    port => 12201
  }
}
output {
  stdout {}
  elasticsearch {
    hosts => ["http://elasticsearch:9200"]
  }
}
```

Finally, launch the components that compose the Elastic Stack:

- Elasticsearch
- Logstash
- Kibana

You can do this via the following docker-compose file that you can launch via `docker-compose run -d`:

```
# Launch Elasticsearch
version: '3.2'

services:
  elasticsearch:

>     image: docker.elastic.co/elasticsearch/elasticsearch-oss:6.8.2

Reviewer: Same here about the Elastic stack version.

    ports:
      - "9200:9200"
      - "9300:9300"
    environment:
      ES_JAVA_OPTS: "-Xms512m -Xmx512m"
    networks:
      - elk

  logstash:
    image: docker.elastic.co/logstash/logstash-oss:6.8.2
    volumes:
      - source: $HOME/pipelines
        target: /usr/share/logstash/pipeline
        type: bind
    ports:
      - "12201:12201/udp"
      - "5000:5000"
      - "9600:9600"
    networks:
      - elk
    depends_on:
      - elasticsearch

  kibana:
    image: docker.elastic.co/kibana/kibana-oss:6.8.2
    ports:
      - "5601:5601"
    networks:
      - elk
    depends_on:
      - elasticsearch

networks:
  elk:
    driver: bridge

```

> Launch your application, you should see your logs arriving inside the Elastic Stack; you can use Kibana to access them.

Reviewer: Maybe add a link to the Kibana UI?

Author: done

== Send logs to Fluentd (EFK)

First, you need to create a Fluentd image with the needed plugins: elasticsearch and input-gelf.
You can use the following Dockerfile that should be created inside a `fluentd` directory.

```
FROM fluent/fluentd:v1.3-debian
RUN ["gem", "install", "fluent-plugin-elasticsearch", "--version", "3.7.0"]
RUN ["gem", "install", "fluent-plugin-input-gelf", "--version", "0.3.1"]
```

You can build the image or let docker-compose build it for you.

Then you need to create a fluentd configuration file inside `$HOME/fluentd/fluent.conf`
```
<source>
  type gelf
  tag example.gelf
  bind 0.0.0.0
  port 12201
</source>

<match example.gelf>
  @type elasticsearch
  host elasticsearch
  port 9200
  logstash_format true
</match>

```

Finally, launch the components that compose the EFK Stack:

- Elasticsearch
- Fluentd
- Kibana

You can do this via the following docker-compose file that you can launch via `docker-compose run -d`:
```
version: '3.2'

services:
  elasticsearch:

>     image: docker.elastic.co/elasticsearch/elasticsearch-oss:6.8.2

Reviewer: Same here about Elastic version.

    ports:
      - "9200:9200"
      - "9300:9300"
    environment:
      ES_JAVA_OPTS: "-Xms512m -Xmx512m"
    networks:
      - efk

  fluentd:
    build: fluentd
    ports:
      - "12201:12201/udp"
    volumes:
      - source: $HOME/fluentd
        target: /fluentd/etc
        type: bind
    networks:
      - efk
    depends_on:
      - elasticsearch

  kibana:
    image: docker.elastic.co/kibana/kibana-oss:6.8.2
    ports:
      - "5601:5601"
    networks:
      - efk
    depends_on:
      - elasticsearch

networks:
  efk:
    driver: bridge
```

> Launch your application, you should see your logs arriving inside EFK: you can use Kibana to access them.

Reviewer: Same here about the Kibana URI.

Author: done

== Fluentd alternative: use SYSLOG

> You can also send your logs to Fluentd using a SYSLOG input.

Reviewer: can you do `s/SYSLOG/Syslog/g`?

> As opposed to GELF input, SYSLOG input will not render multiline logs in one event, that's why we advise to use the GELF input that we implement in Quarkus.

Reviewer: ```suggestion
As opposed to the GELF input, the SYSLOG input will not render multiline logs in one event, that's why we advise to use the GELF input that we implement in Quarkus.
```

> First, you need to create a Fluentd image with the needed plugins: elasticsearch and input-gelf.

Reviewer: You don't have `input-gelf` in the code below and it doesn't seem needed.

Author: done

You can use the following Dockerfile that should be created inside a `fluentd` directory.

```
FROM fluent/fluentd:v1.3-debian
RUN ["gem", "install", "fluent-plugin-elasticsearch", "--version", "3.7.0"]
```

Then, you need to create a fluentd configuration file inside `$HOME/fluentd/fluent.conf`
```
<source>
  @type syslog
  port 5140
  bind 0.0.0.0
  message_format rfc5424
  tag system
</source>

<match **>
  @type elasticsearch
  host elasticsearch
  port 9200
  logstash_format true
</match>
```

Then, launch the components that compose the EFK Stack:

- Elasticsearch
- Fluentd
- Kibana

You can do this via the following docker-compose file that you can launch via `docker-compose run -d`:
```
version: '3.2'

services:
  elasticsearch:

>     image: docker.elastic.co/elasticsearch/elasticsearch-oss:6.8.2

Reviewer: Same here about Elastic version.

Maybe it's worth making it a variable of the Asciidoc file?

    ports:
      - "9200:9200"
      - "9300:9300"
    environment:
      ES_JAVA_OPTS: "-Xms512m -Xmx512m"
    networks:
      - efk

  fluentd:
    build: fluentd
    ports:
      - "5140:5140/udp"
    volumes:
      - source: $HOME/fluentd
        target: /fluentd/etc
        type: bind
    networks:
      - efk
    depends_on:
      - elasticsearch

  kibana:
    image: docker.elastic.co/kibana/kibana-oss:6.8.2
    ports:
      - "5601:5601"
    networks:
      - efk
    depends_on:
      - elasticsearch

networks:
  efk:
    driver: bridge
```

Finally, configure your application to send logs to EFK using SYSLOG:
```
quarkus.log.syslog.enable=true
quarkus.log.syslog.endpoint=localhost:5140
quarkus.log.syslog.protocol=udp
quarkus.log.syslog.app-name=quarkus
quarkus.log.syslog.hostname=quarkus-test
```

> Launch your application, you should see your logs arriving inside EFK: you can use Kibana to access them.

Reviewer: Same here about the link to the Kibana UI.

Author: done

[[configuration-reference]]
== Configuration Reference

> Configuration is done through the normal `application.properties` file.

Reviewer: ```suggestion
Configuration is done through the usual `application.properties` file.
```

include::{generated-dir}/config/quarkus-logging-gelf.adoc[opts=optional, leveloffset=+1]

This extension uses the `logstash-gelf` library that allow more configuration options via system properties,

> you can access it's documentation here: https://logging.paluch.biz/

Reviewer: ```suggestion
you can access its documentation here: https://logging.paluch.biz/
```

In this guide, we will expose how to send them to an external tool using the `quarkus-logging-gelf` extension that can use TCP or UDP to send logs in the Graylog Extented Log Format (GELF).

The `quarkus-logging-gelf` extension will add a GELF log handler to the underlying logging backend that Quarkus uses (jboss-logmanager).
By default, it is disabled, if you enable it but still use another handler (by default the console handler is enabled), your logs will be sent to both handlers.

== Example application

The following examples will all be based on the same example application that you can create based on the following steps:

- Create an application with the `quarkus-logging-gelf` extension. You can use the following Maven command to create it:
```
mvn io.quarkus:quarkus-maven-plugin:1.0.0.Final:create \
    -DprojectGroupId=org.acme \
    -DprojectArtifactId=gelf-logging \
    -DclassName="org.acme.quickstart.GelfLoggingResource" \
    -Dpath="/gelf-logging" \
    -Dextensions="logging-gelf"
```

- Add a log inside your rest endpoint
```
@Path("/gelf-logging")
@ApplicationScoped

> public class GelfLoggingResource {

Reviewer 2: The purpose of this endpoint is just to have something that triggers logging, right?  If so, it should probably be made clear that you don't need this endpoint in order to make this work.

Author: I add a comment

    private static final Logger LOG = Logger.getLogger(GelfLogHandlerResource.class);

    @GET
    public void log() {
        LOG.info("Some useful log message");
    }

}
```

- Configure the gelf log handler to send logs to an external UDP endpoint on the port 12201
```
quarkus.log.handler.gelf.enabled=true
quarkus.log.handler.gelf.host=localhost
quarkus.log.handler.gelf.port=12201
```

> WARNING: UDP logging is currently not supported on native mode, use TCP if you plan to deploy your application as a native image.

Reviewer 2: This might not be true anymore now that we're on GraalVM 19.3.0.

Author: On GraalVM 19.3 it didn't works anymore!

```
Exception in thread "main" java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.doStart(ApplicationImpl.zig:191)
	at io.quarkus.runtime.Application.start(Application.java:87)
	at io.quarkus.runtime.Application.run(Application.java:210)
	at io.quarkus.runner.GeneratedMain.main(GeneratedMain.zig:41)
Caused by: java.lang.NullPointerException
	at java.net.InetAddress.getHostFromNameService(InetAddress.java:615)
	at java.net.InetAddress.getCanonicalHostName(InetAddress.java:589)
	at biz.paluch.logging.RuntimeContainer.isQualified(RuntimeContainer.java:148)
	at biz.paluch.logging.RuntimeContainer.getInetAddressWithHostname(RuntimeContainer.java:137)
	at biz.paluch.logging.RuntimeContainer.lookupHostname(RuntimeContainer.java:90)
	at biz.paluch.logging.RuntimeContainer.initialize(RuntimeContainer.java:67)
	at biz.paluch.logging.gelf.jul.GelfLogHandler.<init>(GelfLogHandler.java:54)
	at io.quarkus.logging.gelf.GelfLogHandlerRecorder.initializeHandler(GelfLogHandlerRecorder.java:17)
	at io.quarkus.deployment.steps.GelfLogHandlerProcessor$build28.deploy_0(GelfLogHandlerProcessor$build28.zig:76)
	at io.quarkus.deployment.steps.GelfLogHandlerProcessor$build28.deploy(GelfLogHandlerProcessor$build28.zig:36)
	at io.quarkus.runner.ApplicationImpl.doStart(ApplicationImpl.zig:68)
	... 3 more
```

Reviewer 2: Interesting; any idea where the null reference is coming from?

Author: I use a workaround to configure the hostname resolver to localhost before network and it fixes the issue. 
I propose to open a followup issue to fix this properly after merging this, I suspect and issue on GraalVM side as the NPE is from code inside the JDK : `java.net.InetAddress.getHostFromNameService()`

    environment:
      ES_JAVA_OPTS: "-Xms512m -Xmx512m"
    networks:
      - graylog

  mongo:
    image: mongo:4.0
    networks:
      - graylog

  graylog:
    image: graylog/graylog:3.1
    ports:
      - "9000:9000"
      - "12201:12201/udp"
      - "1514:1514"
    environment:
      GRAYLOG_HTTP_EXTERNAL_URI: "http://127.0.0.1:9000/"
    networks:
      - graylog
    depends_on:
      - elasticsearch
      - mongo

networks:
  graylog:
    driver: bridge
```

Then, you need to create a UDP input in Graylog.
You can do it from the Graylog web console (System -> Input -> Select GELF UDP) or via the API.

This curl example will create a new Input of type GELF UDP, it uses the default login from Graylog (admin/admin).

```
curl -H "Content-Type: application/json" -H "Authorization: Basic YWRtaW46YWRtaW4=" -H "X-Requested-By: curl" -X POST -v -d \
'{"title":"udp input","configuration":{"recv_buffer_size":262144,"bind_address":"0.0.0.0","port":12201,"decompress_size_limit":8388608},"type":"org.graylog2.inputs.gelf.udp.GELFUDPInput","global":true}' \
http://localhost:9000/api/system/inputs
```

Launch your application, you should see your logs arriving inside Graylog.

> NOTE: due to the initialization time of the GELF log handler, it is possible that not all startup logs are sent to Graylog.

Reviewer 2: Because of the way we set up our log handlers, the few "missed" log messages should be automatically replayed.  If this isn't working, let's have a chat to figure out the cause of the problem.

Author: Yes, it's the case. I wasn't sure so I write this note. Will delete it.

Author: deleted

  elasticsearch {
    hosts => ["http://elasticsearch:9200"]
  }
}
```

Finally, launch the components that compose the Elastic Stack:

- Elasticsearch
- Logstash
- Kibana

You can do this via the following docker-compose file that you can launch via `docker-compose run -d`:

```
# Launch Elasticsearch
version: '3.2'

services:
  elasticsearch:
    image: docker.elastic.co/elasticsearch/elasticsearch-oss:6.8.2
    ports:
      - "9200:9200"
      - "9300:9300"
    environment:
      ES_JAVA_OPTS: "-Xms512m -Xmx512m"
    networks:
      - elk

  logstash:
    image: docker.elastic.co/logstash/logstash-oss:6.8.2
    volumes:
      - source: $HOME/pipelines
        target: /usr/share/logstash/pipeline
        type: bind
    ports:
      - "12201:12201/udp"
      - "5000:5000"
      - "9600:9600"
    networks:
      - elk
    depends_on:
      - elasticsearch

  kibana:
    image: docker.elastic.co/kibana/kibana-oss:6.8.2
    ports:
      - "5601:5601"
    networks:
      - elk
    depends_on:
      - elasticsearch

networks:
  elk:
    driver: bridge

```

Launch your application, you should see your logs arriving inside the Elastic Stack; you can use Kibana to access them.

> NOTE: due to the initialization time of the GELF log handler, it is possible that not all startup logs are sent to the Elastic Stack.

Reviewer 2: Here too.

networks:
  graylog:
    driver: bridge
----

Then, you need to create a UDP input in Graylog.
You can do it from the Graylog web console (System -> Input -> Select GELF UDP) or via the API.

This curl example will create a new Input of type GELF UDP, it uses the default login from Graylog (admin/admin).

> [source,shell script]

Reviewer: I think it should be just `shell`.

Author: This is what IDEA proposes.
I try to generate the doc with both and this does not seems to be any difference so I replace by script.
